homeassistant/components/unifiprotect/migrate.py

```diff
@@ -81,3 +86,68 @@ async def async_migrate_buttons(
 
     if count < len(to_migrate):
         _LOGGER.warning("Failed to migate %s reboot buttons", len(to_migrate) - count)
+
+
+async def async_migrate_device_ids(
+    hass: HomeAssistant, entry: ConfigEntry, protect: ProtectApiClient
+) -> None:
+    """
+    Migrate unique IDs from {device_id}_{name} format to {mac}_{name} format.
+
+    This makes devices persist better with in HA. Anything a device is unadopted/readopted or
+    the Protect instance has to rebuild the disk array, the device IDs of Protect devices
+    can change. This causes a ton of orphaned entities and loss of historical data. MAC
+    addresses are the one persistent identifier a device has that does not change.
+
+    Added in 2022.7.0.
+    """
+
+    registry = er.async_get(hass)
+    to_migrate = []
+    for entity in er.async_entries_for_config_entry(registry, entry.entry_id):
+        parts = entity.unique_id.split("_")
+        # device ID = 24 characters, MAC = 12
+        if len(parts[0]) == 24:
+            _LOGGER.debug("Entity %s needs migration", entity.entity_id)
+            to_migrate.append(entity)
+
+    if len(to_migrate) == 0:
+        _LOGGER.debug("No entities need migration to MAC address ID")
+        return
+
+    try:
+        bootstrap = await protect.get_bootstrap()
+    except (TimeoutError, ClientError, ServerDisconnectedError) as err:
+        raise ConfigEntryNotReady from err
```

Comment:
Maybe move this block into its own function since its the exact same as `async_migrate_buttons` then we only have 1 uncovered line instead of 2